tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/NN.java

```diff
@@ -0,0 +1,379 @@
+              Collections.singletonList(
+                  Constant.scalarOf(
+                      scope,
+                      "Shape mismatch: The shape of labels  "
+                          + "should equal the shape of logits except for the last "
+                          + "dimension "))));
+    }
+
+    // Reshape logits to 2 dim, labels to 1 dim.
+    long numClassses = logitsShape.size(logitsShape.numDimensions() - 1);
+
+    precise_logits =
+        Reshape.create(
+            scope, precise_logits, Constant.vectorOf(scope, new long[] {-1, numClassses}));
+    labels = Reshape.create(scope, labels, Constant.scalarOf(scope, -1));
+    scope.withControlDependencies(shapeChecks);
+    SparseSoftmaxCrossEntropyWithLogits smax =
+        SparseSoftmaxCrossEntropyWithLogits.create(scope, precise_logits, labels);
+    Operand cost = smax.loss();
+    cost = Reshape.create(scope, cost, labelsShape);
+    if (logits.asOutput().dataType() == TFloat16.DTYPE) {
+      cost = Cast.create(scope, cost, TFloat16.DTYPE);
+    }
+    return cost;
+  }
+
+  /**
+   * Computes sigmoid cross entropy given `logits`.
+   *
+   * <p>Measures the probability error in discrete classification tasks in which each class is
+   * independent and not mutually exclusive. For instance, one could perform multilabel
+   * classification where a picture can contain both an elephant and a dog at the same time.
+   *
+   * <p>For brevity, let `x = logits`, `z = labels`. The logistic loss is
+   *
+   * <pre>
+   *     z * -log(sigmoid(x)) + (1 - z) * -log(1 - sigmoid(x))
+   *     = z * -log(1 / (1 + exp(-x))) + (1 - z) * -log(exp(-x) / (1 + exp(-x)))
+   *     = z * log(1 + exp(-x)) + (1 - z) * (-log(exp(-x)) + log(1 + exp(-x)))
+   *     = z * log(1 + exp(-x)) + (1 - z) * (x + log(1 + exp(-x))
+   *     = (1 - z) * x + log(1 + exp(-x))
+   *     = x - x * z + log(1 + exp(-x))
+   * </pre>
+   *
+   * <p>For x < 0, to avoid overflow in exp(-x), we reformulate the above
+   *
+   * <pre>
+   *      x - x * z + log(1 + exp(-x))
+   *      = log(exp(x)) - x * z + log(1 + exp(-x))
+   *      = - x * z + log(1 + exp(x))
+   * </pre>
+   *
+   * <p>Hence, to ensure stability and avoid overflow, the implementation uses this equivalent
+   * formulation
+   *
+   * <pre>
+   *     max(x, 0) - x * z + log(1 + exp(-abs(x)))
+   * </pre>
+   *
+   * <p>`logits` and `labels` must have the same type and shape.
+   *
+   * @param scope The TensorFlow scope
+   * @param labels the labels
+   * @param logits the logits of type float32 or float64
+   * @param <T> the type of labels and logits
+   * @return the component-wise logistic losses.
+   */
+  @Endpoint(name = "sigmoidCrossEntropyWithLogits")
+  public static <T extends TNumber> Operand<T> sigmoidCrossEntropyWithLogits(
+      Scope scope, Operand<T> labels, Operand<T> logits) {
+    if (labels.asOutput().shape().numDimensions() != logits.asOutput().shape().numDimensions())
```

Comment:
[Missing brackets](https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used) for `if` block

Reply:
OK, I actually changed it to:
```
if (!isCompatible(labels.asOutput().shape(), logits.asOutput().shape())) {
      throw new IllegalArgumentException(
          String.format(
              "logits and labels must have the same shape (%s vs %s)",
              labels.asOutput().shape().toString(), logits.asOutput().shape()));
    }
``` 
and added method `isCompatible`.
```
 shapes are compatible if they have the same number of dimensions, and
 if the corresponding dimensions are equal, or at least one of the corresponding dimensions is unknown
```

Should `isCompatible` be a part of Shape.java?